addons/suggest_locations/models/stock_move_line.py

```diff
@@ -0,0 +1,132 @@
+# -*- coding: utf-8 -*-
+from collections import defaultdict
+from odoo import models, fields, api, _
+from odoo.exceptions import ValidationError
+from .suggest_locations_policy import SUGGEST_LOCATION_REGISTRY
+
+NON_DROPABLE_STATES = ("draft", "done", "cancel")
+CONTRAINTS_REQUIRING_A_CHECK = ("enforce", "enforce_with_empty")
+WITH_EMPTY_LOCATIONS = ("suggest_with_empty", "enforce_with_empty")
+VIEW_SET = set(["view"])
+
+
+class StockMoveLine(models.Model):
+    _inherit = "stock.move.line"
+
+    def write(self, values):
+        """Extend write to catch preprocessed location update as constrains
+        happens after write creating a check which can never fail
+        Example: If product A has suggested locations 01, 02, but we bypass constraints
+        and write to 03. When we validate it later, with constraints, then no error
+        is raised, even though it reaches to the raise Validation error stage.
+        We allow for a write to be done on a location that is a view, and only validate 
+        if it's location has already been set to a non-view location for all
+        policies that have preprocessing enabled.
+        """
+        Location = self.env["stock.location"]
+
+        if "location_dest_id" in values:
+            locations = Location.browse(values["location_dest_id"])
+            self.filtered(
+                lambda ml: ml.location_dest_id.usage not in VIEW_SET
+            ).validate_location_dest(locations=locations)
+
+        return super(StockMoveLine, self).write(values)
+
+    def _get_policy_class(self, picking_type):
+        """Get the policy for suggesting locations
+        Note: Optionally we could cache the results with @lru_cache to avoid always returning the
+        policy class, but this caused issued with closed cursors. To avoid this,
+        you need to update the environment at the same time you return the cached result.
+        Possibly by having another function that returns the cached class and updated environment
+        by policy.env = self.env
+        """
+        try:
+            policy_type = picking_type.u_suggest_locations_policy
+            policy_class = SUGGEST_LOCATION_REGISTRY[policy_type]
+            return policy_class(self.env)
+        except KeyError:
+            raise ValueError(_("Policy with name=%s could not be found") % policy_type)
+
+    def suggest_locations(self, picking=None, values=None, limit=30):
+        """Suggest locations for move line, either by self or picking_type and values"""
+        # Validate preconditions
+        if not self and (not picking or not values):
+            raise ValueError(
+                _(
+                    "Missing information! If suggest locations is not done via self, picking "
```

Comment:
I think we can simplify this message to something like: Missing information to suggest locations, please provide either move lines or picking and values.